src/tools/cephfs/DataScan.cc

```diff
@@ -0,0 +1,1267 @@
+          return -EINVAL;
+        }
+      } else {
+        std::cerr << "Unknown argument '" << arg << "'" << std::endl;
+        return -EINVAL;
+      }
+    }
+  }
+
+  // Default to output to metadata pool
+  if (driver == NULL) {
+    driver = new MetadataDriver();
+  }
+
+  dout(4) << "connecting to RADOS..." << dendl;
+  rados.connect();
+  r = driver->init(rados, mdsmap);
+  if (r < 0) {
+    return r;
+  }
+
+  std::string const &command = args[0];
+
+  // Initialize data_io for those commands that need it
+  if (command == "scan_inodes"
+     || command == "scan_extents") {
+    if (args.size() < 2) {
+      usage();
+      return -EINVAL;
+    }
+
+    const std::string data_pool_name = args[args.size() - 1];
+    {
+      data_pool_id = rados.pool_lookup(data_pool_name.c_str());
+      if (data_pool_id < 0) {
+        std::cerr << "Data pool '" << data_pool_name << "' not found!" << std::endl;
+        return -ENOENT;
+      } else {
+        dout(4) << "data pool '" << data_pool_name
+          << "' has ID " << data_pool_id << dendl;
+      }
+
+      if (!mdsmap->is_data_pool(data_pool_id)) {
+        std::cerr << "Warning: pool '" << data_pool_name << "' is not a "
+          "CephFS data pool!" << std::endl;
+      }
+
+      dout(4) << "opening data pool '" << data_pool_name << "'" << dendl;
+      r = rados.ioctx_create(data_pool_name.c_str(), data_io);
+      if (r != 0) {
+        return r;
+      }
+    }
+  }
+
+  // Finally, dispatch command
+  if (command == "scan_inodes") {
+    return recover();
+  } else if (command == "scan_extents") {
+    return recover_extents();
+  } else if (command == "init") {
+    return driver->init_roots(mdsmap->get_first_data_pool());
+  } else {
+    std::cerr << "Unknown command '" << command << "'" << std::endl;
+    return -EINVAL;
+  }
+}
+
+int MetadataDriver::inject_unlinked_inode(
+    inodeno_t inono, int mode, int64_t data_pool_id)
+{
+  // Compose
+  object_t oid = InodeStore::get_object_name(inono, frag_t(), ".inode");
+  InodeStore inode;
+  inode.inode.ino = inono;
+  inode.inode.version = 1;
+  inode.inode.xattr_version = 1;
+  inode.inode.mode = 0500 | mode;
+  // Fake size to 1, so that the directory doesn't appear to be empty
+  // (we won't actually give the *correct* size here though)
+  inode.inode.size = 1;
+  inode.inode.dirstat.nfiles = 1;
+
+  inode.inode.ctime = 
+    inode.inode.mtime = ceph_clock_now(g_ceph_context);
+  inode.inode.nlink = 1;
+  inode.inode.truncate_size = -1ull;
+  inode.inode.truncate_seq = 1;
+
+  // Force layout to default: should we let users override this so that
+  // they don't have to mount the filesystem to correct it?
+  inode.inode.layout = g_default_file_layout;
+  inode.inode.layout.fl_pg_pool = data_pool_id;
+
+  // Serialize
+  bufferlist inode_bl;
+  ::encode(std::string(CEPH_FS_ONDISK_MAGIC), inode_bl);
+  inode.encode(inode_bl);
+
+  // Write
+  int r = metadata_io.write_full(oid.name, inode_bl);
```

Comment:
Maybe we should check the object doesn't actually exist? (With *sigh* an option to delete and override if we know it's bad.)